extensions/amp-form/0.1/amp-form.js

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {user} from '../../../src/log';
+import {getService} from '../../../src/service';
+import {xhrFor} from '../../../src/xhr';
+import {childElement, removeElement} from '../../../src/dom';
+import {templatesFor} from '../../../src/template';
+import {toArray} from '../../../src/types';
+import {onDocumentReady} from '../../../src/document-ready';
+import {isExperimentOn} from '../../../src/experiments';
+
+/** @const @type {string} */
+const TAG = 'amp-form';
+
+class AmpForm {
+
+  constructor(element, win) {
+    /** @const @private {!Window} */
+    this.win_ = win;
+
+    /** @const @private {!Element} */
+    this.form_ = element;
+
+    /** @const @private {!Templates} */
+    this.templates_ = templatesFor(this.win_);
+
+    /** @const @private {!Xhr} */
+    this.xhr_ = xhrFor(this.win_);
+
+    /** @const @private {string} */
+    this.method_ = this.form_.getAttribute('method') || 'POST';
+
+    /** @const @private {string} */
+    this.action_ = user.assert(this.form_.getAttribute('action'),
+        'amp-form requires action attribute %s', this.form_);
+
+    this.installSubmitHandler_();
+  }
+
+  /** @private */
+  installSubmitHandler_() {
+    this.form_.addEventListener('submit', e => {
+      // Safari does not trigger validation check on submission, hence we
+      // trigger it manually. In other browsers this would never execute since
+      // the submit event wouldn't be fired if the form is invalid.
+      if (!this.form_.checkValidity()) {
+        this.form_.classList.remove('valid');
```

Comment:
Ditto on class names.

Why "valid"/"invalid" class names? Why not `:invalid` and `:valid`?